lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp

```diff
@@ -781,75 +782,144 @@ SILCombiner::visitInjectEnumAddrInst(InjectEnumAddrInst *IEAI) {
+    assert(InitEnumBB && "DataAddrInst is not in a valid Basic Block");
+    assert(InjectEnumBB && "IEAI is not in a valid Basic Block");
+    llvm::SmallVector<SILBasicBlock *, 64> Worklist;
+    Worklist.push_back(InjectEnumBB);
+    llvm::SmallSetVector<SILBasicBlock *, 16> Preds;
+    Preds.insert(InjectEnumBB);
+    bool ReachedInitEnumBB = false;
+    while (!Worklist.empty()) {
+      SILBasicBlock *BB = Worklist.pop_back_val();
+
+      // Check current BB for any of the MayWriteSet instructions
+      // TODO - add inst_range to SILBasicBlock to iterate over range
+      for (auto InsIt = BB->begin(); InsIt != BB->end(); ++InsIt) {
+        SILInstruction *Ins = &*InsIt;
+        if (MayWriteSet.count(Ins) != 0) {
+          return nullptr;
+        }
+      }
+
+      if (BB == InitEnumBB) {
+        ReachedInitEnumBB = true;
+        continue;
+      }
+
+      // Go to predecessors and do all that again
+      for (SILBasicBlock *Pred : BB->getPreds()) {
+        // If it's already in the set, then we've already queued and/or
+        // processed the predecessors.
+        if (Preds.insert(Pred)) {
+          if (ReachedInitEnumBB) {
+            // InitEnumBB does not dominate InjectEnumBB
+            return nullptr;
+          }
+          Worklist.push_back(Pred);
+        }
+      }
+    }
+  }
+
   if (SI) {
+    assert((SI->getDest() == DataAddrInst) &&
+           "Can't find StoreInst with DataAddrInst as its destination");
     // In that case, create the payload enum/store.
-    EnumInst *E =
-      Builder.createEnum(DataAddrInst->getLoc(), SI->getSrc(),
-                         DataAddrInst->getElement(),
-                         DataAddrInst->getOperand()->getType().getObjectType());
+    EnumInst *E = Builder.createEnum(
+        DataAddrInst->getLoc(), SI->getSrc(), DataAddrInst->getElement(),
+        DataAddrInst->getOperand()->getType().getObjectType());
     Builder.createStore(DataAddrInst->getLoc(), E, DataAddrInst->getOperand());
     // Cleanup.
     eraseInstFromFunction(*SI);
     eraseInstFromFunction(*DataAddrInst);
     return eraseInstFromFunction(*IEAI);
   }
 
+  // Check whether we have an apply initializing the enum.
+  //  %iedai = init_enum_data_addr %enum_addr
+  //         = apply(%iedai,...)
+  //  inject_enum_addr %enum_addr
+  //
+  // We can localize the store to an alloc_stack.
+  // Allowing us to perform the same optimization as for the store.
+  //
+  //  %alloca = alloc_stack
+  //            apply(%alloca,...)
+  //  %load = load %alloca
+  //  %1 = enum $EnumType, $EnumType.case, %load
+  //  store %1 to %nopayload_addr
+  //
   assert(AI && "Must have an apply");
+  unsigned ArgIdx = 0;
+  for (auto &Opd : AI->getArgumentOperands()) {
+    // Found an apply that initializes the enum. We can optimize this by
+    // localizing the initialization to an alloc_stack and loading from it.
+    DataAddrInst = dyn_cast<InitEnumDataAddrInst>(Opd.get());
```

Comment:
Since we need the argument index, we'd need to create an EnumeratorRange to use in the predicate. for such a small loop the resulting code is larger...